CMakeLists.txt

```diff
@@ -204,33 +224,180 @@ ELSE()
+
+  # ETI infrastructure to be used by components
+  include(cmake/kokkoskernels_eti.cmake)
+  SET(ETI_HEADERS)
+
+  LIST(APPEND HEADERS ${CMAKE_CURRENT_BINARY_DIR}/${PACKAGE_NAME}_config.h)
+
+  # Only enable requested components
+  # if nothing requested, build everything
+  # to recover original behavior.
+  INCLUDE(common/CMakeLists.txt)
+  IF (KokkosKernels_ENABLE_BATCHED)
+    INCLUDE(batched/CMakeLists.txt)
+  ENDIF()
+  IF (KokkosKernels_ENABLE_BLAS)
+    INCLUDE(blas/CMakeLists.txt)
+  ENDIF()
+  IF (KokkosKernels_ENABLE_GRAPH)
+    INCLUDE(graph/CMakeLists.txt)
+  ENDIF()
+  IF (KokkosKernels_ENABLE_SPARSE)
+    INCLUDE(sparse/CMakeLists.txt)
+  ENDIF()
+
+  FOREACH(DIR ${KK_INCLUDE_DIRS})
+    KOKKOSKERNELS_INCLUDE_DIRECTORIES(${DIR})
+    APPEND_GLOB(HEADERS ${DIR}/*.hpp)
+  ENDFOREACH()
+  # MESSAGE("HEADERS: ${HEADERS}")
+
+
+  # For now, don't add the ETI headers to complete list of headers
+  # This will break some IDEs and won't properly display headers in the file list
+  # However, because of the behavior of TRIBITS_ADD_LIBRARY
+  # I can't add the ETI headers with them getting double-installed
+  # with a bunch of headers in the wrong location
+  # This doesn't change pre-existing behavior before the ETI changes
+  #LIST(APPEND HEADERS ${ETI_HEADERS})
+  #-----------------------------------------------------------------------------
+
+  KOKKOSKERNELS_ADD_LIBRARY(
+    kokkoskernels
+    HEADERS ${HEADERS}
+    SOURCES ${SOURCES}
+  )
+
+  # It does not seem to be possible to automatically have headers installed
+  # in the correct subdirectory structure using HEADERS and ADD_LIBRARY with tribits
+  # Force the specializations to be installed in a subdirectory
+  INSTALL(FILES ${ETI_HEADERS} DESTINATION ${KOKKOSKERNELS_HEADER_INSTALL_DIR}/generated_specializations_hpp/)
+
+  IF (KOKKOSKERNELS_HAS_TRILINOS)
+    #no linking commands required - tribits does this
+  ELSE()
+    ADD_LIBRARY(Kokkos::kokkoskernels ALIAS kokkoskernels)
+    TARGET_LINK_LIBRARIES(kokkoskernels PUBLIC Kokkos::kokkos)
+    FOREACH(DIR ${KK_INCLUDE_DIRS})
+      TARGET_INCLUDE_DIRECTORIES(kokkoskernels PUBLIC $<BUILD_INTERFACE:${DIR}>)
+    ENDFOREACH()
+    TARGET_INCLUDE_DIRECTORIES(kokkoskernels PUBLIC
+      $<INSTALL_INTERFACE:${KOKKOSKERNELS_HEADER_INSTALL_DIR}>)
+  ENDIF()
+
+  IF (KOKKOS_ENABLE_SYCL)
+    SET(KOKKOSKERNELS_INTEL_ARCHS ${Kokkos_ARCH})
+    LIST(FILTER KOKKOSKERNELS_INTEL_ARCHS INCLUDE REGEX ".*INTEL.*")
+    LIST(LENGTH KOKKOSKERNELS_INTEL_ARCHS KOKKOSKERNELS_INTEL_ARCHS_NUM)
+    IF(KOKKOSKERNELS_INTEL_ARCHS_NUM GREATER_EQUAL 1)
+      IF (NOT BUILD_SHARED_LIBS)
+        MESSAGE(SEND_ERROR
+          "At the moment, we require KokkosKernels (and Kokkos) to be built as "
+          "shared libraries to allow querying free and total device memory!"
+        )
+      ENDIF()
+    TARGET_LINK_LIBRARIES(kokkoskernels PUBLIC ze_loader)
+    ENDIF()
+  ENDIF()
+
+  KOKKOSKERNELS_LINK_TPL(kokkoskernels PUBLIC LAPACK)
```

Comment:
Could we have a CMake list of TPLs and loop over them to call `KOKKOSKERNELS_LINK_TPL` on each of them one by one instead of those copy/paste?

Reply:
We could but it's out of scope for this PR, I do not know the rational for that decision, I did read a few times that the CMake folks at Kitware recommend copy/pasting over loops. My guess is that it is done to avoid generating complex string manipulation in CMake but I'm not really sure...